### PR TITLE
HmGui styling

### PR DESCRIPTION
**Pull request by HaronK**

### CODE REVIEW

File: engine/lib/phx/src/ui/hmgui/README.md
Comment (Flatfingers) [inline engine/lib/phx/src/ui/hmgui/README.md:1]: Suggested changes (separated by "----"):

From:

prop.string: Test

to:

prop.string: "Test"

----

From:

This feature is used for automatic property value transferring from one element to another. Let's take a `button` element as an example. Inside, it contains `text` element that expects `text.font` and `text.color` properties. They cannot be set directly for the button, so we register `button.text-font` and `button.text-color`, and map them to the corresponding `text` properties. Imagine we have defined `button` element in Lua:

to:

This feature is used for automatically transferring a property value from one element to another. Let's take a `button` element as an example. The `button` element internally contains a `text` element that expects `text.font` and `text.color` properties. To preserve the benefits of Immediate Mode UI definition, properties cannot be set directly on an element. Instead, we define ("register") special variables that can be accessed from LuaJIT, and map these variables to the various properties of elements.

In our `button` example, we first register `button.text-font` and `button.text-color`, then we map these to the corresponding `text` properties of `button`. Imagine we have defined a `button` element in Lua:

----

From:

### Property methods

- `Gui:getPropertyType(id)`: returns property type
- `Gui:mapProperty(id)`: copies property value to it's mapped properties for the current following element. Should be used inside element function definition.
- `Gui:removeProperty(id)`: remove property from the current element style
- `Gui:registerProperty*(name, value, map_id*)`: register a new property with optional id of the mapped property
- `Gui:setProperty*(id, value)`: set property value
- `Gui:getProperty*(id)`: get property value

to:

### Property methods

HmGui provides several methods allowing Lua scripters to manage element properties:

- `Gui:getPropertyType(id)`: returns property type
- `Gui:mapProperty(id)`: copies property value to its mapped properties for the current following element. Should be used inside element function definition.
- `Gui:removeProperty(id)`: remove property from the current element style
- `Gui:registerProperty(name, value, map_id)`: register a new property with optional id of the mapped property
- `Gui:setProperty(id, value)`: set property value
- `Gui:getProperty(id)`: get property value

[Note: I removed the "*" from registerProperty, map_id, setProperty, and getProperty on the assumption that they were copied from the Rust code but would not appear in Lua calls to these methods. If you were using the "*" for some other reason, let's define it.]

----

From:

1. Default properties. Those are hardcoded in the engine in [core_properties.rs](core_properties.rs) file. These properties are used if property was not found in other 2 groups. There is an autogenerated enum containing ids of all core properties: `GuiProperties`.
2. Themes. Theme is a big set of properties that overhauls the default ones. Theme should be defined in a separate Yaml file in the `theme` resource folder. Name of the file will be used as the theme name. All themes are automatically loaded by the engine during startup.
Methods to work with themes in the scripts:
   - `Gui:setTheme(name)`: to set theme
   - `Gui:clearTheme()`: remove theme and fall back to the default properties
Setting a theme is a heavy operation and should not be used too often, i.e. every frame.
Theme property value is used if it was not found in the last group.
3. Element styles. This type of style is used for per element styling. There are 2 ways to do it:
  - explicitly in the code. Example:
```lua
Gui:clearStyle()
Gui:setPropertyVec4(GuiProperties.ButtonTextColorId, Vec4(1, 0, 0, 1))
Gui:button("MyButton")
```
  - via style configuration file. During startup the engine loads all element styles from the `styles.yaml` configuration file in the resource folder. [Example](../../../test_data/styles.yaml). Usage:
```lua
Gui:setStyle(Enums.Gui.Styles.MyButtonStyleId)
Gui:button("MyButton")
```
Ids of the element styles should be initialized in the [`GuiEnums.lua`](/script/Enums/GuiEnums.lua) file.

to:

1. **Default properties**. These are hardcoded in the engine in the [core_properties.rs](core_properties.rs) file. These default property values are used if a value for a required property was not found in the other two groups. An autogenerated `GuiProperties` enum contains the IDs of all core properties.

2. **Themes**. A "theme" is a collection (possibly a large collection!) of element properties that replace the default property values. Specific named themes are defined in individual Yaml files in the `\theme` resource folder. The name of the file will be used as the theme name. All themes are automatically loaded by the engine during startup.

The following methods can be used in Lua scripts to manage themes:
   - `Gui:setTheme(name)`: to set theme
   - `Gui:clearTheme()`: remove theme and fall back to the default properties
Setting a theme is a heavy operation and should not be used too often, i.e. every frame.
Any property value applied by setting a theme that contains that property will be used only if it is not overridden by a per-element styling on that property (see group 3 below).

3. **Per-element styles**. This type of styling sets a property on an individual element. There are 2 ways to do it: either explictly in Lua code, or in a global style configuration file.

  - Example of direct styling in code:
```lua
Gui:clearStyle()
Gui:setPropertyVec4(GuiProperties.ButtonTextColorId, Vec4(1, 0, 0, 1))
Gui:button("MyButton")
```
  - Example of styling through the global style configuration file `styles.yaml`:
```lua
Gui:setStyle(Enums.Gui.Styles.MyButtonStyleId)
Gui:button("MyButton")
```
During startup the engine loads all element styles from the `styles.yaml` configuration file in the resource folder. [Example](../../../test_data/styles.yaml).
IDs of element styles should be initialized in the [`GuiEnums.lua`](/script/Enums/GuiEnums.lua) file.

----

From:

Scripters can define a custom properties in the [`GuiEnums.lua`](/script/Enums/GuiEnums.lua) file. Example:

to:

Scripters can define custom properties in the [`GuiEnums.lua`](/script/Enums/GuiEnums.lua) file. Example:

Comment (HaronK) [reply engine/lib/phx/src/ui/hmgui/README.md:1]: fixed

File: engine/lib/phx/src/render/mesh.rs
Comment (Flatfingers) [inline engine/lib/phx/src/render/mesh.rs:1]: Thank you for checking for memory leaks.

File: res/styles.yaml
Comment (Flatfingers) [inline res/styles.yaml:1]: Is this a placeholder?

**Review by Flatfingers — APPROVED** [review]

**Review by dgavedissian — APPROVED** [review]
Looks good on the Rust side. I love the cleaned up Resource system.

**Pull request merged**